File: tests/cpp/full_example/full_example_test.cpp
#include "generated/full_example.h"

#include <fstream>
#include <iostream>

#include "termite-yaml.h"
#include "termite.hpp"

using namespace test::name::space;

/**
 * @brief Test if an example model can be constructed, saved to YAML and
 * reloaded again
 *
 * @return An error string on error
 */
std::optional<std::string> test_reload() {
  // Create an instance of the data model
  VersionString version = "1.0.1";
  DefaultValues defaults(
      {{"size", Size({{"w", 10}, {"h", 20}})}, {"state", State::TypeFilled()}});
  GeometryList geometries(
      {Rectangle({{"center", Point({{"x", 15}, {"y", -30}})},
                  {"state", State::TypeEdge(5)}}),
       Circle({{"center", Point({{"x", 0}, {"y", 0}})}, {"radius", 7}})});
  DataModel data_model({{"version", version},
                        {"defaults", defaults},
                        {"geometries", geometries}});

  // Save the data model to yaml
  auto node_out = termite::Node::from_value(data_model);
  auto yaml_out = termite::to_YAML(node_out);
  std::ofstream out_file("data_model.yaml");
  out_file << yaml_out;
  out_file.close();

  // Load it again
  std::ifstream in_file("data_model.yaml");
  auto yaml_in = YAML::Load(in_file);
  in_file.close();
  auto node_in = termite::from_YAML(yaml_in);
  if (!node_in.is_ok()) {
    std::stringstream ss;
    ss << "Error loading yaml: " << node_in.get_err();
    return ss.str();
  }
  auto data_model_in = node_in.get_ok().to_value<DataModel>();
  if (!data_model_in.is_ok()) {
    std::stringstream ss;
    ss << "Error converting to DataModel: " << data_model_in.get_err();
    return ss.str();
  }
  auto data_model_loaded = data_model_in.get_ok();
  if (data_model != data_model_loaded) {
    std::stringstream ss;
    ss << "Data model loaded does not match the original, "
       << "Original: " << data_model << ", Loaded: " << data_model_loaded;
    return ss.str();
  }

  return std::nullopt;
}

int main() {
  auto names = {
      "test_reload",
  };
  auto functions = {
      test_reload,
  };

  std::cout << "Running " << names.size() << " tests" << std::endl;

  int progress = 1;
  auto name_it = names.begin();
  for (auto function_it = functions.begin(); function_it < functions.end();
       ++function_it, ++name_it, ++progress) {
    if (auto error = (*function_it)()) {
      std::cout << "Error occured at \"" << *name_it << "\": " << *error
                << std::endl;
      return progress;
    }
  }

  std::cout << "No errors" << std::endl;

  return 0;
}

File: tests/dart/full_example/full_example_test.dart
import 'generated/full_example.dart' as model;
import 'generated/termite-json.dart' as termite_json;
import 'generated/termite-yaml.dart' as termite_yaml;

typedef TestFunction = String? Function();

int runTests(Map<String, TestFunction> tests) {
  print('Running ${tests.length} tests');
  var progress = 1;
  for (final entry in tests.entries) {
    final error = entry.value();
    if (error != null) {
      print('Error occurred at "${entry.key}": $error');
      return progress;
    }
    progress += 1;
  }
  print('No errors');
  return 0;
}

String? testReloadThroughJson() {
  final version = model.VersionString('1.0.1');
  final defaultState = model.State.newEdge(model.SizeValue(1));
  final defaults = model.DefaultValues(
    state: defaultState,
    size: model.Size(w: model.SizeValue(10), h: model.SizeValue(20)),
  );

  final rectangle = model.Rectangle(
    center: model.Point(x: 15, y: -30),
    size: null,
    state: model.State.newEdge(model.SizeValue(5)),
  );
  final circle = model.Circle(
    center: model.Point(x: 0, y: 0),
    radius: model.SizeValue(7),
    state: null,
  );

  final dataModel = model.DataModel(
    version: version,
    defaults: defaults,
    geometries: model.GeometryList([
      model.Geometry.newRectangle(rectangle),
      model.Geometry.newCircle(circle),
    ]),
  );

  final json = termite_json.toString(dataModel.toNode());
  if (!json.isOk()) {
    return 'Failed to serialize full example to JSON: ${json.asError().getMessage()}';
  }

  final parsedNode = termite_json.fromString(json.getOk());
  if (!parsedNode.isOk()) {
    return 'Failed to parse JSON string back into node: ${parsedNode.asError().getMessage()}';
  }

  final loaded = model.DataModel.fromNode(parsedNode.getOk());
  if (!loaded.isOk()) {
    return 'Failed to parse DataModel from node: ${loaded.asError().getMessage()}';
  }

  final loadedModel = loaded.getOk();
  if (loadedModel != dataModel) {
    return 'Model mismatch after reload: $loadedModel';
  }
  return null;
}

String? testReloadThroughYaml() {
  final version = model.VersionString('1.0.1');
  final defaultState = model.State.newEdge(model.SizeValue(1));
  final defaults = model.DefaultValues(
    state: defaultState,
    size: model.Size(w: model.SizeValue(10), h: model.SizeValue(20)),
  );

  final rectangle = model.Rectangle(
    center: model.Point(x: 15, y: -30),
    size: null,
    state: model.State.newEdge(model.SizeValue(5)),
  );
  final circle = model.Circle(
    center: model.Point(x: 0, y: 0),
    radius: model.SizeValue(7),
    state: null,
  );

  final dataModel = model.DataModel(
    version: version,
    defaults: defaults,
    geometries: model.GeometryList([
      model.Geometry.newRectangle(rectangle),
      model.Geometry.newCircle(circle),
    ]),
  );

  final yaml = termite_yaml.toString(dataModel.toNode());
  if (!yaml.isOk()) {
    return 'Failed to serialize full example to YAML: ${yaml.asError().getMessage()}';
  }

  final parsedNode = termite_yaml.fromString(yaml.getOk());
  if (!parsedNode.isOk()) {
    return 'Failed to parse YAML string back into node: ${parsedNode.asError().getMessage()}';
  }

  final loaded = model.DataModel.fromNode(parsedNode.getOk());
  if (!loaded.isOk()) {
    return 'Failed to parse DataModel from node: ${loaded.asError().getMessage()}';
  }

  final loadedModel = loaded.getOk();
  if (loadedModel != dataModel) {
    return 'Model mismatch after reload: $loadedModel';
  }
  return null;
}

void main() {
  final code = runTests({
    'testReloadThroughJson': testReloadThroughJson,
    'testReloadThroughYaml': testReloadThroughYaml,
  });
  if (code != 0) {
    throw Exception('test failure code: $code');
  }
}
